python/pylibraft/pylibraft/neighbors/ivf_pq.pyx

```diff
@@ -0,0 +1,652 @@
+            fast shared memory kernels can be used even for datasets with large dimansionality.
+            Note that the recall is slightly degraded when low-precision type is selected.
+            Possible values [CUDA_R_32F, CUDA_R_16F, CUDA_R_8U]
+        internal_distance_dtype: default = ivf_q.CUDA_R_32F (float)
+            Storage data type for distance/similarity computation.
+            Possible values [CUDA_R_32F, CUDA_R_16F]
+        
+        """
+
+        self.params.n_probes = n_probes
+        self.params.lut_dtype = lut_dtype
+        self.params.internal_distance_dtype = internal_distance_dtype
+        # self.params.shmem_carveout = self.shmem_carveout # TODO(tfeher): enable if #926 adds this
+
+    @property
+    def n_probes(self):
+        return self.params.n_probes
+
+    @property
+    def lut_dtype(self):
+        return self.params.lut_dtype
+
+    @property
+    def internal_distance_dtype(self):
+        return self.params.internal_distance_dtype       
+
+
+@auto_sync_handle
+def search(SearchParams search_params, 
+           Index index, 
+           queries,
+           k,
+           neighbors, 
+           distances,
+           DeviceMemoryResource memory_resource=None,
+           handle=None):
+    """
+    Find the k nearest neighbors for each query.
+
+    Parameters
+    ----------
+    search_params : SearchParams
+    index : Index
+        Trained IVF-PQ index.
+    queries : CUDA array interface compliant matrix shape (n_samples, dim)
+        Supported dtype [float, int8, uint8]
+    k : int
+        The number of neighbors.
+    neighbors : CUDA array interface compliant matrix shape (n_queries, k), dtype uint64_t
+        The neighbor indices will be returned here. 
+    distances : CUDA array interface compliant matrix shape (n_queries, k)
+            The distances to the neighbors will be returned here.
+    memory_resource : RMM DeviceMemoryResource object, optional
+        This can be used to explecitly manage the temporary memory allocation during search. 
+        Passing a pooling allocator can reduce memory allocation overhead. If not specified,
+        then the memory resource from the raft handle is used.
+    {handle_docstring}
+
+    Examples
+    --------
+    .. code-block:: python
+
+        import cupy as cp
+
+        from pylibraft.common import Handle
+        from pylibraft.neighbors import ivf_pq
+
+        n_samples = 50000
+        n_features = 50
+        n_queries = 1000
+        dataset = cp.random.random_sample((n_samples, n_features), dtype=cp.float32)
+
+        # Build index
+        handle = Handle()
+        index = ivf_pq.build(ivf_pq.IndexParams(), dataset, handle=handle)
+
+        # Search using the built index
+        queries = cp.random.random_sample((n_queries, n_features), dtype=cp.float32)
+        k = 10
+        neighbors = cp.empty((n_queries, k), dtype=cp.uint64)
```

Comment:
I'm hoping that after https://github.com/rapidsai/raft/pull/1008 is merged, we can make the explicit output arguments optional (eg setting them to None by default) so that this function will return a tuple or device_ndarray in the default case (indices, distances).

Reply:
We can do that change later, though, and don't need to hold up this PR for it.